knot.x should return non-zero exit code if config file is missing

Currently knot.x returns 0 exit code regardless of config file presence

```
$ java -jar knotx-standalone-1.1.0.fat.jar -conf /path/to/not/exisitng/file.conf; echo $?
12:24:19,036 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback.groovy]
12:24:19,036 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
12:24:19,036 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [jar:file:/opt/knotx/publish/knotx-standalone-1.1.0.fat.jar!/logback.xml]
12:24:19,052 |-INFO in ch.qos.logback.core.joran.spi.ConfigurationWatchList@7225790e - URL [jar:file:/opt/knotx/publish/knotx-standalone-1.1.0.fat.jar!/logback.xml] is not of type file
12:24:19,097 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
12:24:19,102 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
12:24:19,108 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT]
12:24:19,166 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - This appender no longer admits a layout as a sub-component, set an encoder instead.
12:24:19,166 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - To ensure compatibility, wrapping your layout in LayoutWrappingEncoder.
12:24:19,166 |-WARN in ch.qos.logback.core.ConsoleAppender[STDOUT] - See also http://logback.qos.ch/codes.html#layoutInsteadOfEncoder for details
12:24:19,168 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [io.knotx] to INFO
12:24:19,168 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting additivity of logger [io.knotx] to false
12:24:19,169 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[io.knotx]
12:24:19,169 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to ERROR
12:24:19,169 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[ROOT]
12:24:19,169 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
12:24:19,170 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@54a097cc - Registering current configuration as safe fallback point

2017-07-25 12:24:19 [main] ERROR i.v.c.i.launcher.commands.RunCommand - The -conf option does not point to an existing file or is not a valid JSON object
io.vertx.core.json.DecodeException: Failed to decode:Unexpected character ('p' (code 112)): was expecting either '*' or '/' for a comment
 at [Source: /path/to/not/exisitng/file.conf; line: 1, column: 3]
	at io.vertx.core.json.Json.decodeValue(Json.java:82)
	at io.vertx.core.json.JsonObject.fromJson(JsonObject.java:930)
	at io.vertx.core.json.JsonObject.<init>(JsonObject.java:53)
	at io.vertx.core.impl.launcher.commands.RunCommand.getConfiguration(RunCommand.java:418)
	at io.vertx.core.impl.launcher.commands.RunCommand.run(RunCommand.java:247)
	at io.vertx.core.impl.launcher.VertxCommandLauncher.execute(VertxCommandLauncher.java:230)
	at io.vertx.core.impl.launcher.VertxCommandLauncher.dispatch(VertxCommandLauncher.java:384)
	at io.vertx.core.impl.launcher.VertxCommandLauncher.dispatch(VertxCommandLauncher.java:328)
	at io.knotx.launcher.LogbackLauncher.main(LogbackLauncher.java:25)
Caused by: com.fasterxml.jackson.core.JsonParseException: Unexpected character ('p' (code 112)): was expecting either '*' or '/' for a comment
 at [Source: /path/to/not/exisitng/file.conf; line: 1, column: 3]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1586)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:521)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportUnexpectedChar(ParserMinimalBase.java:450)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._skipComment(ReaderBasedJsonParser.java:2358)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._skipWSOrEnd2(ReaderBasedJsonParser.java:2321)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._skipWSOrEnd(ReaderBasedJsonParser.java:2272)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:601)
	at com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:3847)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3792)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2797)
	at io.vertx.core.json.Json.decodeValue(Json.java:79)
	... 8 more
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@                                  ,,,,,,,,,                                 @@
@@                                *,,,,,,,,,,,*                               @@
@@                              @@&,,,,,,,,,,,,,*                             @@
@@                            @@@@@@,,,,,,,,,,,,,,*                           @@
@@                          @@@@@@@@@,,,,,,,,,,,,,,,*                         @@
@@                        @@@@@@@@@@@@/,,,,,,,,,,,,,,,*                       @@
@@                      @@@@@@@@@@@@@@@#,,,,,,,,,,,,,,,,*                     @@
@@                    &@@@@@@@@@@@@@@@@@&,,,,,,,,,,,,,,*@@#                   @@
@@                  @@&,@@@@@@@@@@@@@@@@@@,,,,,,,,,,,,(@@@@@#                 @@
@@                @@@@&,,%@@@@@@@@@@@@@@@@@*,,,,,,,,,%@@@@@@@@#               @@
@@              @@@@@@&,,,*@@@@@@@@@@@@@@@@@(,,,,,,,@@@@@@@@@@@@#             @@
@@            @@@@@@@@&,,,,,@@@@@@@@@@@@@@@@@&,,,,,@@@@@@@@@@@@@@@#           @@
@@          @@@@@@@@@@&,,,,,,%@@@@@@@@@@@@@@@@@,,/@@@@@@@@@@@@@@@@@,*         @@
@@        @@@@@@@@@@@@&,,,,,,,*@@@@@@@@@@@@@@@@@%@@@@@@@@@@@@@@@@#,,,,*       @@
@@      ,@@@@@@@@@@@@@&,,,,,,,,,@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@*,,,,,,,*     @@
@@    ,,,@@@@@@@@@@@@@&,,,,,,,,,,#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@,,,,,,,,,,,*   @@
@@  *,,,,#@@@@@@@@@@@@&,,,,,,,,,,,*@@@@@@@@@@@@@@@@@@@@@@@@@@(,,,,,,,,,,,,,,* @@
@@ ,,,,,,*@@@@@@@@@@@@&,,,,,,,,,,,,,@@@@@@@@@@@@@@@@@@@@@@@@,,,,,,,,,,,,,,,,,*@@
@@,,,,,,,,@@@@@@@@@@@@&,,,,,,,,,,,,,,(@@@@@@@@@@@@@@@@@@@@&,,,,,,,,,,,,,,,,,,,@@
@@/,,,,,,,&@@@@@@@@@@@&,,,,,,,,,,,,,,,*@@@@@@@@@@@@@@@@@@@/,,,,,,,,,,,,,,,,,,,@@
@@ *,,,,,,(@@@@@@@@@@@&,,,,,,,,,,,,,,#@@@@@@@@@@@@@@@@@@@@@&,,,,,,,,,,,,,,,,, @@
@@   ,,,,,*@@@@@@@@@@@&,,,,,,,,,,,,,@@@@@@@@@@@@@@@@@@@@@@@@@,,,,,,,,,,,,,,*  @@
@@     ,,,,@@@@@@@@@@@&,,,,,,,,,,,*@@@@@@@@@@@@@@@@@@@@@@@@@@@(,,,,,,,,,,*    @@
@@       ,,,,,,,,,,,,,,,,,,,,,,,,%@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@,,,,,,,*      @@
@@         ,,&@@@@@@@%,,,,,,,,,,@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@*,,,*        @@
@@           @@@@@@@@@@*,,,,,,/@@@@@@@@@@@@@@@@@*@@@@@@@@@@@@@@@@@#*          @@
@@             @@@@@@@@@,,,,,&@@@@@@@@@@@@@@@@#,,,@@@@@@@@@@@@@@@@            @@
@@               @@@@@@@,,,,@@@@@@@@@@@@@@@@@*,,,,,%@@@@@@@@@@@@              @@
@@                 @@@@*,,(@@@@@@@@@@@@@@@@@,,,,,,,,/@@@@@@@@@                @@
@@                   @,,,&@@@@@@@@@@@@@@@@&,,,,,,,,,,,@@@@@@                  @@
@@                     *@@@@@@@@@@@@@@@@@(,,,,,,,,,,,,,&@@                    @@
@@                       @@@@@@@@@@@@@@@*,,,,,,,,,,,,,,*                      @@
@@                         @@@@@@@@@@@@,,,,,,,,,,,,,,*                        @@
@@                           @@@@@@@@%,,,,,,,,,,,,,*                          @@
@@                             @@@@@/,,,,,,,,,,,,*                            @@
@@                               @@,,,,,,,,,,,,*                              @@
@@                                 *,,,,,,,,,*                                @@
@@                                    ,,,*/                                   @@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

java.lang.NullPointerException: iterable must not be null
	at rx.internal.operators.OnSubscribeFromIterable.<init>(OnSubscribeFromIterable.java:40)
	at rx.Observable.from(Observable.java:1996)
	at io.knotx.launcher.KnotxStarterVerticle.start(KnotxStarterVerticle.java:36)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:434)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:445)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
2017-07-25 12:24:19 [vert.x-eventloop-thread-1] ERROR i.v.c.i.l.c.VertxIsolatedDeployer - Failed in deploying verticle
java.lang.NullPointerException: iterable must not be null
	at rx.internal.operators.OnSubscribeFromIterable.<init>(OnSubscribeFromIterable.java:40)
	at rx.Observable.from(Observable.java:1996)
	at io.knotx.launcher.KnotxStarterVerticle.start(KnotxStarterVerticle.java:36)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:434)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:445)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
java.lang.NullPointerException: iterable must not be null
	at rx.internal.operators.OnSubscribeFromIterable.<init>(OnSubscribeFromIterable.java:40)
	at rx.Observable.from(Observable.java:1996)
	at io.knotx.launcher.KnotxStarterVerticle.start(KnotxStarterVerticle.java:36)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:434)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:445)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,7 +2,8 @@
 All notable changes to Knot.x will be documented in this file.
 
 ## Unreleased
-
+ - [PR-318](https://github.com/Cognifide/knotx/pull/318) - Knot.x returns exit code `30` in case of missing config
+ 
 ## Version 1.1.1
 List of changes that are finished but not yet released in any final version.
  - [PR-316](https://github.com/Cognifide/knotx/pull/316) - Gateway Processor has access to request body

--- a/documentation/src/main/wiki/KnotxDeployment.md
+++ b/documentation/src/main/wiki/KnotxDeployment.md
@@ -42,7 +42,7 @@ java -Dlogback.configurationFile=logback.xml -cp "app/*" io.knotx.launcher.Logba
 
 The execution of Knot.x using a launcher as above it uses a following exit codes as specified in [Vert.x documentation|http://vertx.io/docs/vertx-core/java/#_launcher_and_exit_code].
 Additionally, Knot.x adds following exit codes:
-- `20` - If the configuration is missing or it's empty
+- `30` - If the configuration is missing or it's empty
 
 ### Vert.x metrics
 You might want to enable Vert.x metrics in order to monitor how Knot.x performs. 

--- a/documentation/src/main/wiki/KnotxDeployment.md
+++ b/documentation/src/main/wiki/KnotxDeployment.md
@@ -40,6 +40,10 @@ To start Knot.x with custom modules, use following command
 java -Dlogback.configurationFile=logback.xml -cp "app/*" io.knotx.launcher.LogbackLauncher -conf knotx-starter.json
 ```
 
+The execution of Knot.x using a launcher as above it uses a following exit codes as specified in [Vert.x documentation|http://vertx.io/docs/vertx-core/java/#_launcher_and_exit_code].
+Additionally, Knot.x adds following exit codes:
+- `20` - If the configuration is missing or it's empty
+
 ### Vert.x metrics
 You might want to enable Vert.x metrics in order to monitor how Knot.x performs. 
 Currently, it's possible to enable JMX metrics, so you can use any JMX tool, like JConsole, to inspect all the metrics Vert.x collects.

--- a/knotx-core/src/main/java/io/knotx/launcher/LogbackLauncher.java
+++ b/knotx-core/src/main/java/io/knotx/launcher/LogbackLauncher.java
@@ -16,12 +16,26 @@
 package io.knotx.launcher;
 
 import io.vertx.core.Launcher;
+import io.vertx.core.impl.launcher.commands.ExecUtils;
+import io.vertx.core.json.JsonObject;
 
 public class LogbackLauncher extends Launcher {
+
+  /**
+   * Error code used when the knotx configuration is missing or it's empty
+   */
+  public static final int KNOTX_MISSING_CONFIGURATION_EXIT_CODE = 30;
 
   public static void main(String[] args) {
     System.setProperty("vertx.logger-delegate-factory-class-name",
         "io.vertx.core.logging.SLF4JLogDelegateFactory");
     new LogbackLauncher().dispatch(args);
+  }
+
+  @Override
+  public void afterConfigParsed(JsonObject config) {
+    if (config.isEmpty()) {
+      ExecUtils.exit(KNOTX_MISSING_CONFIGURATION_EXIT_CODE);
+    }
   }
 }

--- a/knotx-core/src/main/java/io/knotx/launcher/LogbackLauncher.java
+++ b/knotx-core/src/main/java/io/knotx/launcher/LogbackLauncher.java
@@ -24,7 +24,7 @@ public class LogbackLauncher extends Launcher {
   /**
    * Error code used when the knotx configuration is missing or it's empty
    */
-  public static final int KNOTX_MISSING_CONFIGURATION_EXIT_CODE = 30;
+  public static final int KNOTX_MISSING_OR_EMPTY_CONFIGURATION_EXIT_CODE = 30;
 
   public static void main(String[] args) {
     System.setProperty("vertx.logger-delegate-factory-class-name",
@@ -35,7 +35,7 @@ public class LogbackLauncher extends Launcher {
   @Override
   public void afterConfigParsed(JsonObject config) {
     if (config.isEmpty()) {
-      ExecUtils.exit(KNOTX_MISSING_CONFIGURATION_EXIT_CODE);
+      ExecUtils.exit(KNOTX_MISSING_OR_EMPTY_CONFIGURATION_EXIT_CODE);
     }
   }
 }

